06 - Type Ahead/index-START.html

```diff
@@ -15,8 +15,68 @@
     </ul>
   </form>
 <script>
-const endpoint = 'https://gist.githubusercontent.com/Miserlou/c5cd8364bf9b2420bb29/raw/2bf258763cdddd704f8ffd3ea9a3e81d25e2c6f6/cities.json';
+  //// global variables
+  const endpoint = 'https://gist.githubusercontent.com/Miserlou/c5cd8364bf9b2420bb29/raw/2bf258763cdddd704f8ffd3ea9a3e81d25e2c6f6/cities.json';
+  let cities = [];
+  let searchTerm = '';
 
+  //// selectors
+  const form = document.querySelector('.search-form');
+  const input = document.querySelector('.search');
+  const suggestions = document.querySelector('.suggestions');
+
+  //// callbacks
+  // fetchData() is async if we decide to also fetch on change or on submit for example
+  const fetchData = async (URL) => {
+    try {
+      const res = await fetch(URL);
+      const data = await res.json();
+      cities = [...data];
+    } catch (error) {
+      const isError = () => {
+        const message = `<h2 class="errorMessage">${error}</h2>`;
+        suggestions.innerHTML = message;  
+      }
+      isError();
+    }
+  }
+
+  fetchData(endpoint);
+
+  const findMatch = (wordToSearch, dataToFilter) => {
+    return dataToFilter.filter(item => {
+      const regex = new RegExp(wordToSearch, 'gi');
+      return item.city.match(regex) || item.state.match(regex);
```

Comment:
have you worked with regex before?

Reply:
Just briefly and I hate it! But I'm aware that it is a super powerful tool especially in this situations. I tried with includes() and it did not work unless you normalise the searchTerm, city, state... too many passages and messy code unless you make utility functions for each of the normalisation steps etc... (just my personal opinion)

Reply:
yeah. using regex are a super nice thing. maybe put it on the list what to practice when you are a mid-level developer :)